docs/docs/reference/metaprogramming/quote-matching.md

```diff
@@ -0,0 +1,64 @@
+---
+layout: doc-page
+title: "Quote Matching Spec"
+---
+
+## Evaluation sematics
+
+Bellow we provide an informal description of the pattern evailatuion.
+We use `'{..}` for expression, `'[..]` for types and `{{..}}` for patterns nested in expressions.
+The semantics are defined as a list of reduction rules that are tried one by one until one matches.
+
+Operations:
+* `s =?= p` checks if a scrutinee `s` matches the pattern `p` while accumulating extracted parts of the code.
+* `isColosedUnder(x1, .., xn)('{e})` returns true if and only if all the references in `e` to names defined in the patttern are contained in the set `{x1, ... xn}`.
+* `lift(x1, .., xn)('{e})` returns `(y1, ..., yn) => [xi = $yi]'{e}` where `yi` is an `Expr` of the type of `xi`.
+* `withEnv(x1 -> y1, ..., xn -> yn)(matching)` evaluates mathing recording that `xi` is equivalent to `yi`.
+* `matched` denotes that the the match succedded and `matched('{e})` denotes that a matech succeded and extracts `'{e}`
+* `&&&` matches if both sides match. Concatenates the extracted expressions of both sides.
+
+```scala
+'{ e } =?= '{ $x }  &&  typeOf('{e}) <:< typeOf('{$x}) && isColosedUnder()('{e})  ===>   matched('{e})
+'{ e } =?= '{ $y(x1, ..., xn) }  &&  isColosedUnder(x1, ... xn)('{e})  ===>   matched(lift(x1, ..., xn)('{e}))
+
+'{ lit } =?= '{ lit }   ===>   matched
+'{ e: T } =?= '{ p: P }   ===>   '{e} =?= '{p} &&& '[T] =?= '[P]
+'{ e } =?= '{ p: P }   ===>   '{e} =?= '{p}
+'{ e.x } =?= '{ p.x }   ===>   '{e} =?= '{p}
+'{ x } =?= '{ x }   ===>   matched
+'{e0.f(e1, ..., en)} =?= '{p0.f(p1, ..., p2)}   ===>   '{e0} =?= '{p0} &&& '{e1} =?= '{p1} &&& ... %% '{en} =?= '{pn}
+'{e.f[T1, ..., Tn]} =?= '{p.f[P1, ..., Pn]}   ===>   '{e} =?= '{p} &&& '[T1] =?= '{P1} &&& ... %% '[Tn] =?= '[Pn]
```

Comment:
It seems we miss a case for `e.f[T1, ..., Tn](e1, ..., en)`, which subsumes the two cases above.

Reply:
Multiple choices are possible, but it seems these two cases match respectively `Apply` and `TypeApply` — and since those are distinct constructors in Tasty, they should be distinct cases in the semantics for consistency, so that each clause matches exactly one term constructor.

Reply:
The problem is that the `e0.f(e1, ..., en)` rule should really be `e0(e1, ..., en)` or else it disallows expressions of the form `x.f[...](e1, ..., en)`.

Reply:
Indeed, that is what the implementation does